xds/src/main/java/io/grpc/xds/client/ControlPlaneClient.java

```diff
@@ -224,6 +225,25 @@ void readyHandler() {
     xdsClient.startSubscriberTimersIfNeeded(serverInfo);
   }
 
+  /**
+   * Indicates whether there is an active ADS stream.
+   *
+   * <p>Return {@code true} when the {@code AdsStream} is created.
+   * {@code false} when the ADS stream fails without a response. Resets to true
+   * upon receiving the first response on a new ADS stream.
+   */
+  // Must be synchronized
+  boolean hasWorkingAdsStream() {
```

Comment:
I think this method should just be `return !streamClosedNoResponse`. `shutdown` shouldn't matter, since this method should not be called when shutdown. `if (adsStream == null) return true` is the same as `if (adsStream == null) return !streamClosedNoResponse` because of earlier checks. And `responseReceived == !streamClosedNoResponse` when `adsStream.closed == true`. When `adsStream.closed == false`, it seems `!streamClosedNoResponse` is more correct than `responseReceived`, as we don't want to incorrectly say it is working.